internal/graph/singleflight_resolver_test.go

```diff
@@ -0,0 +1,371 @@
+			StoreID:              storeID,
+			AuthorizationModelID: model.GetId(),
+			TupleKey:             tupleKey,
+			ResolutionMetadata: &ResolutionMetadata{
+				Depth: 25,
+			},
+		})
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+	}
+}
+
+// TestSingleflightCheckResolver_ContextCancellation is a test to assert
+// that if two FGA queries that have overlapping nested subproblems share
+// different contexts, and one is cancelled out-of-band from the other, then
+// they should both resolve with non-nil errors. If they do not, then that
+// indicates incorrect handling of context cancellation in the call chain.
+func TestSingleflightCheckResolver_ContextCancellation(t *testing.T) {
+	storeID := ulid.Make().String()
+
+	ds := memory.New()
+	err := ds.Write(context.Background(), storeID, nil, []*openfgav1.TupleKey{
+		tuple.NewTupleKey("folder:2", "folder_reader", "user:anne"),
+		tuple.NewTupleKey("folder:2", "allowed", "folder:1"),
+		tuple.NewTupleKey("folder:4", "parent", "folder:2"),
+		tuple.NewTupleKey("folder:4", "allowed", "user:anne"),
+	})
+	require.NoError(t, err)
+
+	model := testutils.MustTransformDSLToProtoWithID(`model
+  schema 1.1
+
+type user
+type folder
+relations
+	define parent: [folder]
+	define folder_reader: [user] or folder_reader from parent
+	define blocked: [user]
+	define allowed: [user] or allowed from parent
+	define reader: folder_reader and allowed
+ `)
+
+	err = ds.WriteAuthorizationModel(context.Background(), storeID, model)
+	require.NoError(t, err)
+
+	ctx := typesystem.ContextWithTypesystem(
+		context.Background(),
+		typesystem.New(model),
+	)
+	ctx = storage.ContextWithRelationshipTupleReader(ctx, ds)
+
+	singleflightCheckResolver := NewSingleflightCheckResolver()
+	localCheckResolver := NewLocalChecker()
+
+	singleflightCheckResolver.SetDelegate(localCheckResolver)
+	localCheckResolver.SetDelegate(singleflightCheckResolver)
+
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+		singleflightCheckResolver.Close()
+		localCheckResolver.Close()
+	})
+
+	var wg sync.WaitGroup
+
+	for i := 0; i < 100; i++ {
+		wg.Add(1)
+		wg.Add(1)
+
+		var err1 error
+		go func() {
+			defer wg.Done()
+
+			ctx2 := ctx
+
+			_, err1 = singleflightCheckResolver.ResolveCheck(ctx2, &ResolveCheckRequest{
+				StoreID:              storeID,
+				AuthorizationModelID: model.GetId(),
+				TupleKey:             tuple.NewTupleKey("folder:4", "reader", "user:anne"),
+				ResolutionMetadata: &ResolutionMetadata{
+					Depth: 25,
+				},
+			})
+		}()
+
+		var err2 error
+		go func() {
+			defer wg.Done()
+
+			ctx1, cancel := context.WithCancel(ctx)
+
+			_, err2 = singleflightCheckResolver.ResolveCheck(ctx1, &ResolveCheckRequest{
+				StoreID:              storeID,
+				AuthorizationModelID: model.GetId(),
+				TupleKey:             tuple.NewTupleKey("folder:2", "reader", "user:anne"),
+				ResolutionMetadata: &ResolutionMetadata{
+					Depth: 25,
+				},
+			})
+
+			cancel()
```

Comment:
> // TestSingleflightCheckResolver_ContextCancellation is a test to assert
// that if two FGA queries that have overlapping nested subproblems share
// different contexts, and one is cancelled out-of-band from the other

i don't think this setup is doing what you want. what guarantee do you have that when `cancel()` runs, the other goroutine hasn't already finished? you need to send the `cancel()`  after the first request arrives at `singleflight.Do`, but before the other one does.

See https://github.com/janos/singleflight/blob/a7affacaf07488d9b49680e2b3f6a24b12186f32/singleflight_test.go#L192 for inspiration. I don't know how we will be able to achieve a similar test, but this is what we should strive for.